pulsar-broker/src/main/java/org/apache/pulsar/broker/service/schema/SchemaRegistryServiceImpl.java

```diff
@@ -180,57 +182,85 @@ public CompletableFuture<List<CompletableFuture<SchemaAndMetadata>>> getAllSchem
-                        .build();
-
-                long start = this.clock.millis();
-
-                return schemaStorage
-                        .put(schemaId, info.toByteArray(), context)
-                        .whenComplete((__, t) -> {
-                            if (t != null) {
-                                log.error("[{}] Put schema failed", schemaId);
-                                this.stats.recordPutFailed(schemaId);
-                            } else {
-                                if (log.isDebugEnabled()) {
-                                    log.debug("[{}] Put schema finished", schemaId);
-                                }
-                                this.stats.recordPutLatency(schemaId, this.clock.millis() - start);
-                            }
-                        });
-
-            });
-
-        }));
+                    if (schemaVersion != null) {
+                        if (log.isDebugEnabled()) {
+                            log.debug("[{}] Schema is already exists", schemaId);
+                        }
+                        return CompletableFuture.completedFuture(schemaVersion);
+                    }
+                    CompletableFuture<Void> checkCompatibilityFuture = new CompletableFuture<>();
+                    if (schemaAndMetadataList.size() != 0) {
+                        if (isTransitiveStrategy(strategy)) {
+                            checkCompatibilityFuture = checkCompatibilityWithAll(schemaId, schema,
+                                    strategy, schemaAndMetadataList);
+                        } else {
+                            checkCompatibilityFuture = checkCompatibilityWithLatest(schemaId,
+                                    schema, strategy);
+                        }
+                    } else {
+                        checkCompatibilityFuture.complete(null);
+                    }
+                    return checkCompatibilityFuture.thenCompose(v -> {
+                        byte[] context = hashFunction.hashBytes(schema.getData()).asBytes();
+                        SchemaRegistryFormat.SchemaInfo info = SchemaRegistryFormat.SchemaInfo
+                                .newBuilder()
+                                .setType(Functions.convertFromDomainType(schema.getType()))
+                                .setSchema(ByteString.copyFrom(schema.getData()))
+                                .setSchemaId(schemaId)
+                                .setUser(schema.getUser())
+                                .setDeleted(false)
+                                .setTimestamp(clock.millis())
+                                .addAllProps(toPairs(schema.getProps()))
+                                .build();
+                        long start = startTime > 0 ? startTime : this.clock.millis();
+                        CompletableFuture<SchemaVersion> persistentFuture = new CompletableFuture<>();
+                        schemaStorage
```

Comment:
I see the current code is that when finds `BadVersion` then tries again until it succeeds, maybe we can rewrite it like this:

```java
long start = startTime > 0 ? startTime : this.clock.millis();
return schemaStorage.put(schemaId, info.toByteArray(), context).handleAsync((sv, ex) -> {
  if (ex != null){
    return CompletableFuture.completedFuture(sv);
  }
  if (xx instanceof xxx){
    return putSchemaIfAbsent(schemaId, schema, strategy, start, retryTime + 1);
  }
  return CompletableFuture.failedFuture(ex);
});
```

this makes it easier to read.

BTW, This change does not seem to solve the problem to be solved

Reply:
> BTW, This change does not seem to solve the problem to be solved

What does this mean?

Reply:
I know the root cause now, how about this suggestion?

```java
long start = startTime > 0 ? startTime : this.clock.millis();
return schemaStorage.put(schemaId, info.toByteArray(), context).handleAsync((sv, ex) -> {
  if (ex != null){
    return CompletableFuture.completedFuture(sv);
  }
  if (xx instanceof xxx){
    return putSchemaIfAbsent(schemaId, schema, strategy, start, retryTime + 1);
  }
  return CompletableFuture.failedFuture(ex);
});
```